pyvrp/read.py

```diff
@@ -552,3 +552,40 @@ def _allowed2profile(self) -> dict[tuple[int, ...], int]:
                 allowed_clients2profile_idx[clients] = next(profile_idx)
 
         return allowed_clients2profile_idx
+
+
+def _problem_data_from_json(cls, data: dict) -> ProblemData:
+    """
+    Creates a :class:`~pyvrp._pyvrp.ProblemData` instance from a JSON-like
+    dictionary.
+
+    Parameters
+    ----------
+    data : dict
+        Dictionary containing all fields required to instantiate a ProblemData
+        object. The structure must match the ProblemData constructor signature
+        (clients, depots, vehicle_types, distance_matrices, duration_matrices,
+        groups).
+
+    Returns
+    -------
+    ProblemData
+        Data instance constructed from the provided dictionary.
+    """
+    clients = [Client(**client) for client in data["clients"]]
+    depots = [Depot(**depot) for depot in data["depots"]]
+    vehicle_types = [VehicleType(**vt) for vt in data["vehicle_types"]]
+    distance_matrices = [np.array(mat) for mat in data["distance_matrices"]]
+    duration_matrices = [np.array(mat) for mat in data["duration_matrices"]]
+    groups = [ClientGroup(**group) for group in data.get("groups", [])]
+    return ProblemData(
+        clients=clients,
+        depots=depots,
+        vehicle_types=vehicle_types,
+        distance_matrices=distance_matrices,
+        duration_matrices=duration_matrices,
+        groups=groups,
+    )
+
+
+setattr(ProblemData, "from_json", classmethod(_problem_data_from_json))
```

Comment:
I struggle with this monkey patch. I'd rather do one of the following:
- Add the method to the `ProblemData` class, on the C++ side. This'd probably require us to bring in e.g. https://github.com/nlohmann/json to parse a JSON-formatted string.
- Make `from_json` a standalone function. But then I do not immediately know where it needs to go in the PyVRP codebase (should it?).